File: services/go-backend/internal/council/orchestrator.go
package council

import (
	"context"
	"fmt"
	"log"
	"os"
	"strings"
	"sync"
	"time"

	"github.com/regular-life/CouncilAI/go-backend/internal/llm"
	"github.com/regular-life/CouncilAI/go-backend/internal/metrics"
)

type CouncilResult struct {
	FinalAnswer      string            `json:"final_answer"`
	Reasoning        string            `json:"reasoning,omitempty"`
	Confidence       float64           `json:"confidence"`
	Source           string            `json:"source"`
	CandidateAnswers []CandidateAnswer `json:"candidate_answers"`
	PeerReviews      []PeerReview      `json:"peer_reviews,omitempty"`
	Latency          time.Duration     `json:"latency"`
}

type CandidateAnswer struct {
	Answer string `json:"answer"`
	Model  string `json:"model"`
	Error  string `json:"error,omitempty"`
}

type PeerReview struct {
	Reviewer string `json:"reviewer"`
	Review   string `json:"review"`
	Error    string `json:"error,omitempty"`
}

type Orchestrator struct {
	clients        []llm.LLMClient
	chairmanClient llm.LLMClient
	stageTimeout   time.Duration
}

func NewOrchestrator(clients []llm.LLMClient, chairmanClient llm.LLMClient, stageTimeout time.Duration) *Orchestrator {
	if stageTimeout == 0 {
		stageTimeout = 30 * time.Second
	}
	return &Orchestrator{
		clients:        clients,
		chairmanClient: chairmanClient,
		stageTimeout:   stageTimeout,
	}
}

func (o *Orchestrator) Query(ctx context.Context, question string, chunks []string, customPrompt string, skipChairman bool) (*CouncilResult, error) {
	start := time.Now()

	if os.Getenv("MOCK_LLM") == "true" {
		time.Sleep(300 * time.Millisecond) // Simulate a very fast LLM
		return &CouncilResult{
			FinalAnswer: "MOCK RESPONSE: Answer generated locally to preserve API quotas. Original Question: " + question,
			Confidence:  0.99,
			Source:      "mock:council",
			Latency:     time.Since(start),
		}, nil
	}

	prompt := customPrompt
	if prompt == "" {
		contextText := strings.Join(chunks, "\n\n---\n\n")
		prompt = fmt.Sprintf(`Based on the following document excerpts, answer the question.

		Document Excerpts:
		%s

		Question: %s

		Instructions:
		- Answer based ONLY on the provided excerpts
		- If the answer is not in the excerpts, say so clearly
		- Be concise but thorough`, contextText, question)
	}

	log.Printf("[Council] Collecting individual responses from %d models", len(o.clients))
	candidates := o.fanOut(prompt)

	var valid []CandidateAnswer
	for _, c := range candidates {
		if c.Error == "" {
			valid = append(valid, c)
		}
	}
	if len(valid) == 0 {
		return nil, fmt.Errorf("all council members failed to respond")
	}
	if len(valid) == 1 {
		log.Printf("[Council] Only 1 model responded, skipping peer review and chairman")
		return &CouncilResult{
			FinalAnswer:      valid[0].Answer,
			Confidence:       0.5,
			Source:           valid[0].Model + " (single-response)",
			CandidateAnswers: candidates,
			Latency:          time.Since(start),
		}, nil
	}

	log.Printf("[Council] Peer review with %d valid candidates", len(valid))
	reviews := o.peerReview(question, valid)

	successfulReviews := 0
	for _, r := range reviews {
		if r.Error == "" {
			successfulReviews++
		}
	}
	if successfulReviews == 0 {
		log.Printf("[Council] Peer review failed entirely, falling back to best candidate")
		best := pickLongestCandidate(valid)
		return &CouncilResult{
			FinalAnswer:      best.Answer,
			Confidence:       0.6,
			Source:           best.Model + " (peer-review-failed-fallback)",
			CandidateAnswers: candidates,
			PeerReviews:      reviews,
			Latency:          time.Since(start),
		}, nil
	}

	if skipChairman {
		log.Printf("[Council] Skipping chairman (skipChairman=true), using best peer-reviewed candidate")
		best := pickLongestCandidate(valid)
		metrics.CouncilResponseTime.Observe(time.Since(start).Seconds())
		return &CouncilResult{
			FinalAnswer:      best.Answer,
			Confidence:       0.75,
			Source:           best.Model + " (peer-reviewed, no chairman)",
			CandidateAnswers: candidates,
			PeerReviews:      reviews,
			Latency:          time.Since(start),
		}, nil
	}

	log.Printf("[Council] Chairman synthesis")
	chairmanResult, err := o.chairmanSynthesize(question, chunks, valid, reviews)
	if err != nil {
		log.Printf("[Council] Chairman synthesis failed: %v, falling back to best candidate", err)
		best := pickLongestCandidate(valid)
		return &CouncilResult{
			FinalAnswer:      best.Answer,
			Confidence:       0.65,
			Source:           best.Model + " (chairman-failed-fallback)",
			CandidateAnswers: candidates,
			PeerReviews:      reviews,
			Latency:          time.Since(start),
		}, nil
	}

	metrics.CouncilResponseTime.Observe(time.Since(start).Seconds())
	return &CouncilResult{
		FinalAnswer:      chairmanResult.Answer,
		Reasoning:        chairmanResult.Reasoning,
		Confidence:       chairmanResult.Confidence,
		Source:           chairmanResult.Source,
		CandidateAnswers: candidates,
		PeerReviews:      reviews,
		Latency:          time.Since(start),
	}, nil
}

func (o *Orchestrator) fanOut(prompt string) []CandidateAnswer {
	var wg sync.WaitGroup
	results := make([]CandidateAnswer, len(o.clients))

	for i, client := range o.clients {
		wg.Add(1)
		go func(idx int, c llm.LLMClient) {
			defer wg.Done()

			ctx, cancel := context.WithTimeout(context.Background(), o.stageTimeout)
			defer cancel()

			resp, err := c.Generate(ctx, prompt)
			if err != nil {
				log.Printf("[Council] Model %s failed: %v", c.ModelName(), err)
				metrics.LLMFailureCount.Inc()
				results[idx] = CandidateAnswer{Model: c.ModelName(), Error: err.Error()}
				return
			}
			results[idx] = CandidateAnswer{Answer: resp.Answer, Model: resp.Model}
		}(i, client)
	}

	wg.Wait()
	return results
}

func (o *Orchestrator) peerReview(question string, candidates []CandidateAnswer) []PeerReview {
	var wg sync.WaitGroup
	reviews := make([]PeerReview, len(o.clients))

	var anonymized []string
	for i, c := range candidates {
		anonymized = append(anonymized, fmt.Sprintf("=== Response %c ===\n%s", 'A'+i, c.Answer))
	}
	answersBlock := strings.Join(anonymized, "\n\n")

	for i, client := range o.clients {
		wg.Add(1)
		go func(idx int, c llm.LLMClient) {
			defer wg.Done()

			prompt := fmt.Sprintf(`You are reviewing multiple AI-generated answers to this question:

Question: %s

Here are the anonymized responses:

%s

Your task:
1. Evaluate each response for accuracy, completeness, and clarity
2. Rank them from best to worst
3. Explain briefly why you ranked them that way

Format your response as:
RANKING: [best to worst, e.g., "B, A, C"]
REASONING: [1-2 sentence explanation of your ranking]`, question, answersBlock)

			ctx, cancel := context.WithTimeout(context.Background(), o.stageTimeout)
			defer cancel()

			resp, err := c.Generate(ctx, prompt)
			if err != nil {
				log.Printf("[Council] Reviewer %s failed: %v", c.ModelName(), err)
				reviews[idx] = PeerReview{Reviewer: c.ModelName(), Error: err.Error()}
				return
			}
			reviews[idx] = PeerReview{Reviewer: c.ModelName(), Review: resp.Answer}
		}(i, client)
	}

	wg.Wait()
	return reviews
}

func pickLongestCandidate(candidates []CandidateAnswer) CandidateAnswer {
	best := candidates[0]
	for _, c := range candidates[1:] {
		if len(c.Answer) > len(best.Answer) {
			best = c
		}
	}
	return best
}

File: backend/app/council/orchestrator.py
import logging
import time

from app.config import get_settings
from app.council.generator import _openrouter_generate, collect_candidates
from app.council.instrumentation import track_llm_calls
from app.council.reviewer import review_candidates
from app.council.synthesizer import synthesize
from app.metrics.prometheus import councilai_council_response_seconds

logger = logging.getLogger(__name__)
settings = get_settings()


async def run_council(
    question: str,
    chunks: list[str],
    custom_prompt: str = "",
    skip_chairman: bool = False,
) -> dict:
    start = time.perf_counter()
    context_text = "\n\n---\n\n".join(chunks)
    prompt = custom_prompt or (
        "Based on the following document excerpts, answer the question.\n\n"
        f"Document Excerpts:\n{context_text}\n\n"
        f"Question: {question}\n\n"
        "Instructions:\n"
        "- Answer based ONLY on the provided excerpts\n"
        "- If the answer is not in the excerpts, say so clearly\n"
        "- Be concise but thorough"
    )

    with track_llm_calls() as stats:
        if settings.llm_provider == "ollama" and settings.local_llm_fast_mode:
            single = await _openrouter_generate(
                settings.ollama_model,
                prompt,
                stage="generate",
                max_tokens=128,
                temperature=0.0,
            )
            if single.error:
                raise RuntimeError(single.error)
            elapsed = time.perf_counter() - start
            councilai_council_response_seconds.observe(elapsed)
            return {
                "final_answer": single.answer,
                "reasoning": "",
                "confidence": 0.6,
                "source": f"{single.model} (single-local-fast)",
                "candidate_answers": [single.__dict__],
                "peer_reviews": [],
                "latency": elapsed,
                "llm_call_count": 1,
                "llm_total_seconds": stats["total_seconds"],
                "llm_by_stage": dict(stats["by_stage"]),
                "council_mode": "single_local_fast",
            }

        candidates = [candidate.__dict__ for candidate in await collect_candidates(prompt)]
        valid_candidates = [candidate for candidate in candidates if not candidate.get("error")]
        if not valid_candidates:
            raise RuntimeError("all council members failed to respond")
        if len(valid_candidates) == 1:
            elapsed = time.perf_counter() - start
            councilai_council_response_seconds.observe(elapsed)
            return {
                "final_answer": valid_candidates[0]["answer"],
                "reasoning": "",
                "confidence": 0.5,
                "source": f"{valid_candidates[0]['model']} (single-response)",
                "candidate_answers": candidates,
                "peer_reviews": [],
                "latency": elapsed,
                "llm_call_count": stats["count"],
                "llm_total_seconds": stats["total_seconds"],
                "llm_by_stage": dict(stats["by_stage"]),
                "council_mode": "single_response",
            }

        reviews = [review.__dict__ for review in await review_candidates(question, valid_candidates)]
        if skip_chairman:
            best = max(valid_candidates, key=lambda c: len(c["answer"]))
            elapsed = time.perf_counter() - start
            councilai_council_response_seconds.observe(elapsed)
            return {
                "final_answer": best["answer"],
                "reasoning": "",
                "confidence": 0.75,
                "source": f"{best['model']} (peer-reviewed, no chairman)",
                "candidate_answers": candidates,
                "peer_reviews": reviews,
                "latency": elapsed,
                "llm_call_count": stats["count"],
                "llm_total_seconds": stats["total_seconds"],
                "llm_by_stage": dict(stats["by_stage"]),
                "council_mode": "peer_review_no_chairman",
            }

        try:
            chairman = await synthesize(question, chunks, valid_candidates, reviews)
            chairman_error: str | None = None
        except Exception as exc:
            logger.warning(
                "[Council] chairman synthesis failed exc_type=%s exc=%s; "
                "falling back to best valid candidate",
                type(exc).__name__,
                exc,
            )
            best = max(valid_candidates, key=lambda c: len(c["answer"]))
            chairman = {
                "answer": best["answer"],
                "reasoning": "",
                "confidence": 0.65,
                "source": f"{best['model']} (chairman-fallback)",
            }
            chairman_error = f"{type(exc).__name__}: {exc}"

        elapsed = time.perf_counter() - start
        councilai_council_response_seconds.observe(elapsed)
        result = {
            "final_answer": chairman["answer"],
            "reasoning": chairman.get("reasoning", ""),
            "confidence": chairman.get("confidence", 0.7),
            "source": chairman.get("source", "chairman-synthesis"),
            "candidate_answers": candidates,
            "peer_reviews": reviews,
            "latency": elapsed,
            "llm_call_count": stats["count"],
            "llm_total_seconds": stats["total_seconds"],
            "llm_by_stage": dict(stats["by_stage"]),
            "council_mode": "full_council",
        }
        if chairman_error is not None:
            result["chairman_error"] = chairman_error
        return result
